### PR TITLE
`General`: Add GitLab and GitLab CI docker configuration

### PR DESCRIPTION
**Pull request by bensofficial**
#### General
- [x] I tested **all** changes and their related features locally and on a test VM.
- [X] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [X] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).

### Motivation and context
To use Artemis with GitLab and GitLab CI (instead of Jenkins), a Docker Compose file is needed for local setup and testing.

### Description
I have created a docker-compose file and instructions for setting up GitLab and the GitLab Runner (which is very similar to the steps for testing).

### Steps for Testing
Prerequisites:
- Docker and Docker compose

# GitLab and GitLab CI setup

First, configure the environment parameters:
```bash
cp src/main/docker/env.example.gitlab-gitlabci.txt src/main/docker/.env
vi src/main/docker/.env
```

Run the following command to start GitLab and a GitLab Runner in a Docker container:
```bash
docker-compose -f src/main/docker/gitlab-gitlabci.yml --env-file src/main/docker/.env up --build -d
```

Then log on to http://localhost/ with the password (`sudo docker exec -it gitlab grep 'Password:' /etc/gitlab/initial_root_password`) and go to http://localhost/admin/runners.
Click on "Register an instance runner" and copy the registration token.
Open a shell into the `gitlab-runner` container:
```bash
docker exec -it gitlab-runner /bin/bash
```
Execute the following command with your token inside the container with the registration token generated above (If you use Linux, try `--clone-url http://172.17.0.1:80 \`, because `gateway.docker.internal` is only for [Windows](https://docs.docker.com/desktop/windows/networking/#use-cases-and-workarounds)):
```bash
gitlab-runner register \
  --non-interactive \
  --executor "docker" \
  --docker-image alpine:latest \
  --url http://gitlab:80 \
  --clone-url http://gateway.docker.internal:80 \
  --registration-token "PROJECT_REGISTRATION_TOKEN" \
  --description "docker-runner" \
  --maintenance-note "Just a random local test runner" \
  --tag-list "docker,artemis" \
  --run-untagged="true" \
  --locked="false" \
  --access-level="not_protected"
```

Now the runner should be successfully registered.
Create a new repository in GitLab and activate CI/CD (Settings -> General -> Visibility, Project features, Permissions -> CI/CD).
Then go back to the root of the repository and click `Set up CI/CD`. If you do not see the button, you can access it via the URL `http://localhost/USERNAME/REPONAME/-/ci/editor`.
When you click on `Create new CI/CD pipeline`, the editor should open with the default template. You can also call up the template [here](https://gitlab.com/gitlab-org/gitlab/-/blob/master/lib/gitlab/ci/templates/Getting-Started.gitlab-ci.yml).
Commit the changes.
The defined pipeline should now run successfully, and all jobs should pass (see CI/CD -> Jobs or `http://localhost/USERNAME/REPONAME/-/jobs`). 

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [ ] Test 2

### CODE REVIEW

File: src/main/docker/gitlab-gitlabci.yml [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        environment:
+            GITLAB_OMNIBUS_CONFIG: |
+                external_url 'https://EXTERNALURL'
+                prometheus_monitoring['enable'] = false
+                gitlab_rails['gitlab_shell_ssh_port'] = 2222
+                gitlab_rails['monitoring_whitelist'] = ['0.0.0.0/0']
+                letsencrypt['enable'] = true
+                letsencrypt['auto_renew_hour'] = "12"
+                letsencrypt['auto_renew_minute'] = "30"
+                letsencrypt['auto_renew_day_of_month'] = "*/7"
```
Comment (Mtze) [inline src/main/docker/gitlab-gitlabci.yml:19]: I think you should exclude these from the base configuration. Let's encrypt requires the service to be available publicly. For local testing this is not the case. I suggest you add a [docker-compose override](https://docs.docker.com/compose/extends/) for production use and switch off HTTPS for local dev deployments.

File: src/main/docker/gitlab/README.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
+# Local GitLab and GitLab CI setup
+
+First, configure the hostname and external URL in the `gitlab-gitlabci.yml` file and run the following command to start GitLab and a GitLab Runner in a Docker container:
+```bash
+docker-compose -f src/main/docker/gitlab-gitlabci.yml up --build -d
+```
+
+Then log on to https://localhost/ with the password (`sudo docker exec -it NAME_OF_THE_CONTAINER grep 'Password:' /etc/gitlab/initial_root_password`) and go to https://localhost/admin/runners.
+Click on Register an instance runner and copy the registration token.
```
Comment (Mtze) [inline src/main/docker/gitlab/README.md:9]:
```suggestion
Click on "Register an instance runner" and copy the registration token.
```

File: src/main/docker/gitlab-gitlabci.yml [hunk 3]
```diff
@@ -0,0 +1,37 @@
+version: "3.6"
+services:
+    gitlab:
+        build: gitlab
+        volumes:
+            - artemis-gitlab-data:/var/opt/gitlab
+            - artemis-gitlab-logs:/var/log/gitlab
+            - artemis-gitlab-config:/etc/gitlab
+        hostname: 'HOSTNAME'
```
Comment (Mtze) [inline src/main/docker/gitlab-gitlabci.yml:9]: Why did you add this?

File: src/main/docker/gitlab-gitlabci.yml [hunk 4]
```diff
@@ -0,0 +1,37 @@
+version: "3.6"
+services:
+    gitlab:
+        build: gitlab
```
Comment (Mtze) [inline src/main/docker/gitlab-gitlabci.yml:4]:
```suggestion
        build: gitlab
        container_name: gitlab
```

This allows easy referencing (See readme comment)

File: src/main/docker/gitlab-gitlabci.yml [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+                letsencrypt['auto_renew_hour'] = "12"
+                letsencrypt['auto_renew_minute'] = "30"
+                letsencrypt['auto_renew_day_of_month'] = "*/7"
+        ports:
+            - '2222:22'
+            - '80:80'
+            - '443:443'
+        shm_size: '256m'
+    gitlab-runner:
+        image: gitlab/gitlab-runner:latest
```
Comment (Mtze) [inline src/main/docker/gitlab-gitlabci.yml:26]:
```suggestion
        image: gitlab/gitlab-runner:latest
        container_name: gitlab-runner
```

(See above and readme comment)

File: src/main/docker/gitlab/README.md [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+Open a shell into the container:
```
Comment (Mtze) [inline src/main/docker/gitlab/README.md:10]:
```suggestion
Open a shell into the `gitlab-runner` container:
```

File: src/main/docker/gitlab/README.md [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+`````bash
+docker ps
+docker exec -it NAME_OF_THE_CONTAINER /bin/bash
```
Comment (Mtze) [inline src/main/docker/gitlab/README.md:13]:
```suggestion
docker exec -it gitlab-runner /bin/bash
```
This only works, if you adapt the docker-compose file according to my suggestions.

**Review by Mtze — CHANGES_REQUESTED** [review]
I tried to execute the docker-compose locally and it failed due to the missing let's encrypt certificate. Added some inline comments how to resolve that

File: src/main/docker/gitlab-gitlabci.yml (hunk 3 above)
Comment (bensofficial) [reply src/main/docker/gitlab-gitlabci.yml]: Good point. I had a problem with it before, so I added it, but actually you don't need it.

File: src/main/docker/gitlab-gitlabci.yml (hunk 1 above)
Comment (bensofficial) [reply src/main/docker/gitlab-gitlabci.yml]: Thank you! I have just improved it.

**Review by dfuchss — COMMENTED** [review]
How would/do you ensure that the build output of the runner cannot be seen by the students?

I do not get the actual purpose of this PR (at its current state). The CI pipeline seems not to be integrated into Artemis. Or do I do not see the definition?

**Comment by bensofficial** [comment]
@dfuchss Actually, the integration of GitLab CI into Artemis is the topic of my thesis.
I have discussed with @jpbernius and @mtze that I can open a pull request even if it is not implemented in Artemis yet.
So the current status is just the setup for testing further development for using Artemis with GitLab and GitLab CI.

**Comment by jpbernius** [comment]
We have to check if we can configure GitLab to hide CI/CD from students.
Otherwise, there is more future work planned to use Artemis as a proxy for GitLab and not allow students to access the GitLab UI at all.

File: src/main/docker/gitlab-gitlabci.prod.yml [hunk 8]
```diff
@@ -0,0 +1,12 @@
+version: "3.6"
+services:
+    gitlab:
+        environment:
+            GITLAB_OMNIBUS_CONFIG: |
```
[the remark this answers is not in the document]
Comment (Mtze) [reply src/main/docker/gitlab-gitlabci.prod.yml:5]: Maybe "prod" is a bit misleading. It is the config for the "remote teste server". I think it is fine to leave that here for now - the real prod config has to be handled by the Admin team.

**Review by Mtze — APPROVED** [review]
Changes LGTM

Tested the setup locally - work fine

File: src/main/docker/gitlab-gitlabci.prod.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (bensofficial) [reply src/main/docker/gitlab-gitlabci.prod.yml:5]: Yes, exactly. It is for test servers. Would it be better if I name it `gitlab-gitlabci.remotetestserver.yml`.

**Comment by dfuchss** [comment]
> @dfuchss Actually, the integration of GitLab CI into Artemis is the topic of my thesis. I have discussed with @jpbernius and @Mtze that I can open a pull request even if it is not implemented in Artemis yet. So the current status is just the setup for testing further development for using Artemis with GitLab and GitLab CI.

Nice. This sounds interesting. This might reduce the amount of technologies we need to maintain :)

**Review by dfuchss — APPROVED** [review]
Looks good to me. Some minor remarks inlined :)

File: src/main/docker/gitlab-gitlabci.prod.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (bensofficial) [reply src/main/docker/gitlab-gitlabci.prod.yml]: I have added a template for an `.env` file in which you can also configure whether a certificate should be used. This also works well with http only. So that there are no further misunderstandings regarding a `prod.yml` file.

File: src/main/docker/gitlab-gitlabci.yml [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    gitlab:
+        build: gitlab
+        container_name: gitlab
+        volumes:
+            - artemis-gitlab-data:/var/opt/gitlab
+            - artemis-gitlab-logs:/var/log/gitlab
+            - artemis-gitlab-config:/etc/gitlab
+        environment:
+            GITLAB_OMNIBUS_CONFIG: |
+                external_url "http://${GIT_SERVER_NAME}"
```
Comment (dfuchss) [inline src/main/docker/gitlab-gitlabci.yml:12]:
```suggestion
                external_url "${GIT_SERVER_NAME}"
```

Simply add the protocol to the environment variable as well :)

File: src/main/docker/gitlab-gitlabci.yml [hunk 10]
```diff
@@ -0,0 +1,52 @@
+version: "3.6"
+services:
+    gitlab:
+        build: gitlab
+        container_name: gitlab
```
Comment (dfuchss) [inline src/main/docker/gitlab-gitlabci.yml:5]: "http://gitlab" would work without naming the containers, why is that needed ? :)

File: src/main/docker/gitlab-gitlabci.yml [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+                gitlab_rails['monitoring_whitelist'] = ['0.0.0.0/0']
+                letsencrypt['enable'] = ${SSL_ENABLED}
+                letsencrypt['auto_renew_hour'] = "12"
+                letsencrypt['auto_renew_minute'] = "30"
+                letsencrypt['auto_renew_day_of_month'] = "*/7"
+        ports:
+            - '2222:22'
+            - '80:80'
+            - '443:443'
+        shm_size: '256m'
```
Comment (dfuchss) [inline src/main/docker/gitlab-gitlabci.yml:24]: Why do you specify shm_size?

File: src/main/docker/gitlab/README.md [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+```bash
+docker exec -it gitlab-runner /bin/bash
+```
+Execute the following command with your token inside the container with the registration token generated above:
+```bash
+gitlab-runner register \
+  --non-interactive \
+  --executor "docker" \
+  --docker-image alpine:latest \
+  --url http://gateway.docker.internal:80 \
```
Comment (dfuchss) [inline src/main/docker/gitlab/README.md:26]:
```suggestion
  --url http://gitlab:80 \
```
you can use gitlab:80 here (at least that should work or am I missing something?)

File: src/main/docker/gitlab/README.md [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+  --clone-url http://gateway.docker.internal:80 \
```
Comment (dfuchss) [inline src/main/docker/gitlab/README.md:27]:
```suggestion
  --clone-url http://gitlab:80 \
```

File: src/main/docker/env.example.gitlab-gitlabci.txt [hunk 14]
```diff
@@ -0,0 +1,3 @@
+GIT_SERVER_NAME=git.example.com
```
Comment (dfuchss) [inline src/main/docker/env.example.gitlab-gitlabci.txt:1]:
```suggestion
GIT_SERVER_NAME=http://git.example.com
```

See comments below :)

**Review by dfuchss — CHANGES_REQUESTED** [review]
LGTM. Just some minor remarks & questions

File: src/main/docker/gitlab-gitlabci.yml (hunk 11 above)
Comment (bensofficial) [reply src/main/docker/gitlab-gitlabci.yml:24]: To be honest, I just looked at the [GitLab documentation](https://docs.gitlab.com/ee/install/docker.html#install-gitlab-using-docker-compose) and there it was. Why exactly that is specified separately, I don't know either. I would guess that otherwise the [default value](https://docs.docker.com/engine/reference/run/#runtime-constraints-on-resources) (`64m`) would be taken and that might be too little for GitLab.

File: src/main/docker/gitlab/README.md (hunk 12 above)
Comment (bensofficial) [reply src/main/docker/gitlab/README.md:26]: Yes, if you only use http, it works. If you use a certificate, there is the problem that it is only valid for the external URL... But then `http://gateway.docker.internal:80` does not work either.

File: src/main/docker/gitlab-gitlabci.yml (hunk 10 above)
Comment (bensofficial) [reply src/main/docker/gitlab-gitlabci.yml:5]: Yes, for accessing the container from the other container we do not need it.
But later it is necessary to execute a command (`sudo docker exec -it gitlab grep 'Password:' /etc/gitlab/initial_root_password`) in the container. There we have to specify the container name. If I were to remove it, the container name would be something like `docker_gitlab_1`, which is just as functional as the current solution, but just adds extra work to the input.

Comment (dfuchss) [reply src/main/docker/gitlab-gitlabci.yml:5]: Ok :)

File: src/main/docker/gitlab-gitlabci.yml (hunk 11 above)
Comment (dfuchss) [reply src/main/docker/gitlab-gitlabci.yml:24]: Fine for me. We didn't specify the value :)

**Review by dfuchss — APPROVED** [review]
Code looks good to me .

**Comment by dfuchss** [comment]
Did you test that the internal urls work? Maybe the clone url has to be an external url because the docker container is not attached to the internal network :)

**Comment by bensofficial** [comment]
That's right! Now that you mention it, it comes back to me 🙈 It must be the external URL. I just tested it on a test server where I entered a "real" URL. I will fix it right now.

File: src/main/docker/gitlab/README.md (hunk 13 above)
Comment (bensofficial) [reply src/main/docker/gitlab/README.md:27]: As described in the conversation section, the container that is started by the runner and then executes the code is not on the network with the Gitlab instance. Therefore, the URL of the host is needed. For Windows, `gateway.docker.internal` works and for Linux `172.17.0.1`.

**Review by dfuchss — APPROVED** [review]
Reapprove

**Comment by dfuchss** [comment]
@bensofficial https://docs.gitlab.com/ee/user/permissions.html#gitlab-cicd-permissions this are the permissions 
you can set to the ci as far as I can see. Students are developers and after due date they are reporters in their own repositories. It seems to be difficult to hide that from developers , but I'm looking forward to the solution.

**Comment by bensofficial** [comment]
Thank you very much for the information. At the moment we have just checked that it is possible to hide the build plan from students by using a custom CI configuration file (https://docs.gitlab.com/ee/ci/pipelines/settings.html#specify-a-custom-cicd-configuration-file). This configuration for that is in the settings that developers do not have access to.
The primary goal now is to make it work and hiding build results / logs has no impact on functionality.
But yes, we have to see how / if it works (or not).

**Review by Mtze — APPROVED** [review]
LGMT 


Thank you!

**Review by jpbernius — APPROVED** [review]
